File: solutions/115/main.py
from typing import List 

class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        # dp table
        # dp[i][j] means no. of sub of t[0:i-1] for s[0:j-1]
        dp = [[0] * (len(t)+1) for _ in range(len(s)+1)]
        # initialize
        for i in range(1, len(t)+1):
            dp[0][i] = 0
        for i in range(1, len(s)+1):
            dp[i][0] = 1
        dp[0][0] = 1
        # traverse dp table
        for i in range(1, len(s)+1):
            for j in range(1, len(t)+1):
                if s[i-1] == t[j-1]:
                    dp[i][j] = dp[i-1][j] + dp[i-1][j-1]
                else:
                    dp[i][j] = dp[i-1][j]
        return dp[-1][-1]

if __name__ == "__main__": 
    s = "rabbbit"
    t = "rabbit"
    sol = Solution()
    assert sol.numDistinct(s, t) == 3

File: solutions/115/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Fri Nov 26 20:34:36 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <string> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    int numDistinct(string s, string t) {
        // dp table
        vector<vector<int>> dp(s.size()+1, vector<int>(t.size()+1, 0));
        // initialize
        dp[0][0] = 1;
        for (int i = 1; i <= t.size(); ++i)
            dp[0][i] = 0;
        for (int j = 1; j <= s.size(); ++j)
            dp[j][0] = 1;
        // traverse dp table
        for (int i = 1; i <= s.size(); ++i) {
            for (int j = 1; j <= t.size(); ++j) {
                if (s[i-1] == t[j-1]) {
                    // to prevent overflow
                    if (dp[i-1][j-1] < INT_MAX - dp[i-1][j])
                        dp[i][j] = dp[i-1][j-1] + dp[i-1][j];
                }
                else {
                    dp[i][j] = dp[i-1][j];
                }
            }
        }
        return dp[s.size()][t.size()];
    }
};

TEST(Test115, SimpleTest) {
    string s = "rabbbit";
    string t = "rabbit";
    Solution sol; 
    EXPECT_EQ(sol.numDistinct(s, t), 3);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}
